bcdaworker/worker/alr.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/CMSgov/bcda-app/bcda/models/fhir/alr"
+	"github.com/CMSgov/bcda-app/bcda/models/postgres"
+	"github.com/CMSgov/bcda-app/bcda/utils"
+	"github.com/CMSgov/bcda-app/bcdaworker/repository"
+	"github.com/CMSgov/bcda-app/conf"
+	"github.com/google/fhir/go/jsonformat"
+	"github.com/google/fhir/go/proto/google/fhir/proto/stu3/resources_go_proto"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/pborman/uuid"
+	"github.com/sirupsen/logrus"
+)
+
+type AlrWorker interface {
+	ProcessAlrJob(ctx context.Context, jobArgs models.JobAlrEnqueueArgs) error
+	UpdateJobAlrStatus(ctx context.Context, jobID uint, newStatus models.JobStatus) error
+	GetAlrJobStatus(ctx context.Context, jobID uint) (*models.JobStatus, error)
+}
+
+type alrWorker struct {
+	*postgres.AlrRepository
+	FHIR_STAGING_DIR string
+	ndjsonFilename   string
+}
+
+const (
+	sqlFlavor = sqlbuilder.PostgreSQL
+)
+
+func NewAlrWorker(db *sql.DB) *alrWorker {
+	alrR := postgres.NewAlrRepo(db)
+
+	// embed data struct that has method GetAlr
+	worker := &alrWorker{
+		AlrRepository:    alrR,
+		FHIR_STAGING_DIR: "",
+		ndjsonFilename:   "", // Filled in later
+	}
+
+	err := conf.Checkout(worker) // worker is already a reference, no & needed
+	if err != nil {
+		logrus.Fatal("Could not get data from conf for ALR.", err)
+	}
+
+	return worker
+}
+
+func (a *alrWorker) ProcessAlrJob(
+	ctx context.Context,
+	jobArgs models.JobAlrEnqueueArgs,
+) error {
+
+	// Parse the jobAlrEnqueueArgs
+	aco := jobArgs.CMSID
+	id := jobArgs.ID
+	MBIs := jobArgs.MBIs
+	lowerBound := jobArgs.LowerBound
+	upperBound := jobArgs.UpperBound
+
+	// Pull the data from ALR tables (alr & alr_meta)
+	alrModels, err := a.GetAlr(ctx, aco, MBIs, lowerBound, upperBound)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	// Set up IO operation to dump ndjson
+	fileUUID := uuid.New()
+	f, err := os.Create(fmt.Sprintf("%s/%d/%s.ndjson", a.FHIR_STAGING_DIR,
+		id, fileUUID))
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+	w := bufio.NewWriter(f)
+	defer utils.CloseFileAndLogError(f)
+
+	// Serialize data into JSON
+	marshaller, err := jsonformat.NewPrettyMarshaller(jsonformat.STU3)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	// data is what will be sent over the channel to the go routine
+	type data struct {
+		patient      *resources_go_proto.Patient
+		observations []*resources_go_proto.Observation
+	}
+
+	// Creating channels for go routine
+	// c is buffered b/c IO operation is slower than unmarshalling
+	c := make(chan data, 1000) // 1000 rows before blocking
+	result := make(chan error)
+
+	// Go rountine + closure that will write to file
+	// Two channels involved with this closure. One receiving "data", and the
+	// other will send an error is there was any error. The worker thread will
+	// will also use the result channel to wait for the IO to finish.
+	go func(c chan data, w *bufio.Writer, result chan error) {
+		// Receive data through channel
+
+		for i := range c {
```

Comment:
🔥  - Just to confirm, the loop exits when the caller does close(c)?

Reply:
Yes.